### DISCUSSION TITLE
Questions with regard to underscore in identifiers

### DISCUSSION THREAD

**Discussion by cosineblast** [cc65/cc65#2343] [Q&A] [answered]
By default, it seems CC65 adds an underscore to identifiers, so something like 
```c
extern u8 stuff[5];
```
gets translated to something equivalent to 
```assembly
.import _stuff
``` 
in the assembly file.
This got me curious to know why that is the case, does it make the implementation easier, or is it made so that  it is easy to identify symbols which are visible to C-generated files?

**Comment by polluks** [comment]
C99 standard says:
7.1.3 Reserved identifiers
- All identifiers that begin with an underscore and either an uppercase letter or another underscore are always reserved for any use.
- All identifiers that begin with an underscore are always reserved for use as identifiers with file scope in both the ordinary and tag name spaces.

**Reply by cosineblast** [reply]
Oh, that' nice to know, but it didn't help me understand why the C symbols cc65 uses  in the assembly output have an underscore, especially since that C99 rule is related to the visibility of symbols in the C source.

Unless ofc cc65 implement this to make it easier to conform to the c99 standard, which would be really interesting.

**Comment by spiro-trikaliotis** [comment, marked as answer]
> This got me curious to know why that is the case, does it make the implementation easier, or is it made so that it is easy to identify symbols which are visible to C-generated files?

This is called "name mangling" or "name decoration" and is absolutely normal for C compilers:
https://en.wikipedia.org/wiki/Name_mangling

In this case, it is a very easy type of name mangling, creating different namespaces for asm (without a leading underscore)  and C (with a leading underscore).

**Reply by cosineblast** [reply]
Oh, I see. I didn't know C compilers did name mangling as well. This makes sense, thanks!
